### PR TITLE
Honor existing attention mask in tokenzier.pad

### LINKED ISSUE #13914: Data collator ignores the input attention_mask
## Environment info

- `transformers` version: 4.11.3
- Platform: Linux-5.4.104+-x86_64-with-Ubuntu-18.04-bionic
- Python version: 3.7.12
- PyTorch version (GPU?): 1.9.0+cu102 (False)
- Tensorflow version (GPU?): 2.6.0 (False)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: No
- Using distributed or parallel set-up in script?: No

### Who can help

@sgugger
@patrickvonplaten 

## Information

The DataCollators do not respect the attention_mask passed to them.

Here's the context:

I was trying to train a GPT2 model where the input_ids are:
```
<PROMPT TOKEN IDS> + <COMPLETION TOKEN IDS>
```

I want to mask out the completion tokens during training, so I set the attention_mask to:

```
[1] * (no. of prompt tokens) + [0] * (no. of completion tokens)
```

The labels are the same as the input_ids, except that the prompt token IDs are replaced with -100 in order to ignore the predictions for prompt tokens in the loss and only penalize the model on predicting completion tokens.

Here is a toy example of the data that would be passed to the data collator following that format:

```python
features = [
  {"input_ids": [1,2,3,4,5,6], "attention_mask": [1,1,1,1,1,0], "labels": [-100,-100,-100,-100,-100,6]},
  {"input_ids": [1,2,3], "attention_mask": [1,1,0], "labels": [-100,-100,4]},
]
```

1-5 in the first example are the prompt token IDs and 6 is the label token ID.
1 and 2 in the second example are the prompt token IDs and 3 is the label token ID. 

I would expect that DataCollatorForSeq2Seq(features) would return the following attention_mask:

```
[[1,1,1,1,1,0],
 [1,1,0,0,0,0]]
```

i.e., padding the second example to make it the same length as the first example.

Instead it returns:

```
[[1,1,1,1,1,0],
 [1,1,1,0,0,0]]
```

Notice that the attention_mask[1,2] is 1 instead of 0 (3 ones instead of 2 ones at the start of the 2nd example), which unmasks the completion token for the 2nd example.

## To reproduce

See this colab: https://colab.research.google.com/drive/1DcvH01SyfIPBwlzqiUCPaSjWKvkAoD2E#scrollTo=wHAfa4t5sZHg

I've also copied and passed the relevant code below:

```python
import transformers

tokenizer = transformers.GPT2Tokenizer.from_pretrained('distilgpt2')
tokenizer.pad_token = tokenizer.unk_token

features = [
  {"input_ids": [1,2,3,4,5,6], "attention_mask": [1,1,1,1,1,0], "labels": [-100,-100,-100,-100,-100,6]},
  {"input_ids": [1,2,3], "attention_mask": [1,1,0], "labels": [-100,-100,4]},
]
output = transformers.DataCollatorForSeq2Seq(tokenizer)(features)

x = output['attention_mask'][0, :].tolist()
print(x) # [1, 1, 1, 0, 0, 0]
assert x == [1,1,0] + [0] * 3
```

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

In #13914, it was highlighted that the `tokenizer.pad` method erases existing `attention_mask` instead of padding them. While it should definitely create them if they don't exist, it shouldn't erase a mask that is passed.
This PR addresses that and adds a common test.

Fixes #13914

### CODE REVIEW

**Review by LysandreJik — APPROVED** [review]
I think this is a good approach, which, as you mentioned, should be mimicked in other inheriting implementations.

**Review by LysandreJik — APPROVED** [review]
This looks good to me. Thank you for taking care of it! I would love a second pair of eyes checking this out before merging.

**Review by patrickvonplaten — APPROVED** [review]
Looked through all changes carefully and everything looks good to me! Thanks for tackling this

**Pull request merged**
